File: analysis/old/rematlabcode/fcTaskDataPrep_Gili.m
%% upload data

clear all; 

% adjust path to where the data is
files = dir(['/Users/isabelpriv/Documents/Princeton/toDelete/normingStudy/norming_session_three/*/*acqext.json']);


%% organize data for each subjects and compute bonus scores
Nsj = length(files); 

allData = {};
for sj=1:Nsj

    exp_minus = [];
    exp_plus = [];
    pleasM = [];
    anxM = [];
    fearM = [];
    
    pleasP = [];
    anxP = [];
    fearP = [];
    
    gad =[];
    phq = [];
  
    correctLetters = 0;
    
    filename = files(sj).name;
    
    % take data saved at end of json file 
    e = jsondecode(fileread([files(sj).folder filesep files(sj).name]));
    d = e{end}; 
   
   %d.debrief_responses
   
   % add expectancy ratings of CS-
   for i = 1:length(d.expectancy_cs_minus)
       if isempty(d.expectancy_cs_minus{i})
           addnum = NaN;
       else
           addnum = str2num(d.expectancy_cs_minus{i});
       end
       exp_minus = [exp_minus addnum];
   end

   % add expectancy ratings of CS+
   for i = 1:length(d.expectancy_cs_plus)
       if isempty(d.expectancy_cs_plus{i})
           addnum = NaN;
       else
           addnum = str2num(d.expectancy_cs_plus{i});
       end
       exp_plus = [exp_plus addnum];
   end
   
   allData(sj).exp_plus = exp_plus'; 
   allData(sj).exp_minus = exp_minus'; 

   % save information on experiment design
   StimSeq = [];
   for i = 1:length(d.stimulus_image)
       a = d.stimulus_image{i};
       if strcmp(a(32), 'c')
           StimSeq = [StimSeq, 1];
       else
           StimSeq = [StimSeq, 2];
       end
             
   end
   
   us = [];
   for i = 1:length(d.stimulus_audio)
       a = d.stimulus_audio{i};
       if length(a)>4 && strcmp(a(34), 's')
           us = [us, 1];
       else
           us = [us, 0];
       end
             
   end
   
   exp_order = [];
   for i = 1:length(d.trial_type)
       a = d.trial_type{i};
       if strcmp(a, 'fctask-trial')
           exp_order = [exp_order, 3];
       elseif strcmp(a, 'expectancy-rating')
           exp_order = [exp_order, 4];
       end             
   end
    
   % save practice ratings
   practRatings = [];
    for i = 1:length(d.trial_type)
       a = d.trial_type{i};
       
       if strcmp(a, 'expectancy-rating')
           newS = e{i};
           if isfield(newS, 'csRating')
               try
               practRatings = [practRatings, str2num(e{i}.csRating)];
               catch
                   practRatings = [practRatings, NaN];
               end
                   
           end
           
       end
    end
    
           

   try
   j=1;
   exp_order_reorg = exp_order;
   for k = 1:length(exp_order)
       if exp_order(k) == 3
           exp_order_reorg(k) = StimSeq(j);
           j = j+1;
       end
   end
   
   b = find(exp_order == 3);
   catch
      
      fprintf('StimSeq not correct for subject %i',sj)
   end
   
   % add affective ratings
   for i = 1:length(d.affective_cs_plus)
       
       if isempty(d.affective_cs_plus{i})
           addnum = NaN;
       else
           addnum = str2num(d.affective_cs_plus{i});
       end
       pleasP  = [pleasP addnum];
       
       
        if isempty(d.affective_cs_minus{i})
           addnum = NaN;
       else
           addnum = str2num(d.affective_cs_minus{i});
       end
       pleasM  = [pleasM addnum];
       
      
   end
   
   % save information for each subject in struct
   allData(sj).pleasM = pleasM;
   allData(sj).pleasP = pleasP;


   allData(sj).RTM = abs(d.fc_rt(StimSeq==1));
   allData(sj).RTP = abs(d.fc_rt(StimSeq==2));
   allData(sj).RT = abs(d.fc_rt);
   
   allData(sj).StimSeq = StimSeq;
   allData(sj).cs = StimSeq;
   allData(sj).us = us;
   try
   allData(sj).exp_order = exp_order;
   catch 
   end
   
   allData(sj).exp_order_reorg = exp_order_reorg;
   allData(sj).ID = filename(1:24);
   allData(sj).d = d;
   allData(sj).trial_type = d.trial_type;
   allData(sj).practRatings = practRatings;
   
end

exp_minus_mat = [];
exp_plus_mat = [];
pleasM_mat = [];
anxM_mat = [];
fearM_mat = [];
pleasP_mat = [];
anxP_mat = [];
fearP_mat = [];
RTM_mat = [];
RTP_mat = [];
RT_mat = [];

for sj=1:Nsj
    if length(allData(sj).exp_minus) == 20
        exp_minus_mat = [exp_minus_mat, allData(sj).exp_minus];
        exp_plus_mat = [exp_plus_mat, allData(sj).exp_plus];
    else
         fprintf('sj %i not added\n', sj)    
    end  
    
end

for sj=1:Nsj
    
    try
        if length(allData(sj).pleasM) == 6
            
            pleasM_mat = [pleasM_mat; allData(sj).pleasM];
            pleasP_mat = [pleasP_mat; allData(sj).pleasP]; 
            
        end
        
        RTP_mat = [RTP_mat, allData(sj).RTP];
        RTM_mat = [RTM_mat, allData(sj).RTM];
        RT_mat = [RT_mat, allData(sj).RT];

    catch
    fprintf('sj %i not added\n', sj)    
    end
    
 
end

mExpMinus = mean(exp_minus_mat,2);
mExpPlus = mean(exp_plus_mat,2);

mPleasM = mean(pleasM_mat);

mPleasP = mean(pleasP_mat);



mRTP = mean(RTP_mat,2);
mRTM = mean(RTM_mat,2);
mRT = mean(RT_mat,2);



% prep data for plotting
   nTrials = (length(mRTM)+length(mRTP));
  
  
   CSplus = [];
   CSminus = [];
   otherCS = [];
    for i=1:nTrials
        if StimSeq(i) == 2 % double-check if StimEst is the same for everybody. 
            CSplus = [CSplus, mRT(i)];
            CSminus = [CSminus, NaN];
        else  
            CSminus = [CSminus, mRT(i)];
            CSplus = [CSplus, NaN];
            
        end
    end
    CSplusInt = fillmissing(CSplus,'linear');
    hold on
    CSminusInt = fillmissing(CSminus,'linear');
    
   
    plot(CSminusInt(5:end),'LineWidth',2);
    hold on
    plot(CSplusInt(5:end),'LineWidth',2);
    


% plot average expectancy ratings and reaction times
figure;

subplot(1,2,1)
 
plot(1:20, mExpMinus, 'linewidth', 3)
hold on
plot(1:20, mExpPlus,  'linewidth', 3)


subplot(1,2,2)

plot(1:28, mRTM)
hold on 
plot(1:34, mRTP)


% uncomment below if we wanto to plot each individual
% for i = 1:size(exp_minus_mat,2)
% figure
%     plot(1:20, exp_minus_mat(:,i), 'linewidth', 3)
%     hold on
%     plot(1:20, exp_plus_mat(:,i), 'linewidth', 3)
% end


% plot affective ratings
figure

set(gca,'FontSize',22)


plot(1:6, mPleasM, 'linewidth', 3)
hold on
plot(1:6, mPleasP, 'linewidth', 3)
ylim([1 100])
ylabel('Pleasant')

set(findall(gcf,'type','text'),'FontSize',22)

File: analysis/old/fcTaskDataPrep_Gili.py
import os  
import json  
import numpy as np  
import matplotlib.pyplot as plt  
from glob import glob


# adjust path to where the data is  
files = glob('../data/data_jun_20/s*_scream_acq_ext_*[0-9].json')

# files = [os.path.join(dp, f) for dp, dn, filenames in os.walk(data_path) for f in filenames if f.endswith('acqext.json')]  
  
Nsj = len(files)  
  
allData = [{} for _ in range(Nsj)]  
  
for sj in range(Nsj):  
    exp_minus = []  
    exp_plus = []  
    pleasM = []  
    anxM = []  
    fearM = []  
  
    pleasP = []  
    anxP = []  
    fearP = []  
  
    gad = []  
    phq = []  
  
    correctLetters = 0  
  
    filename = files[sj]  
  
    # take data saved at end of json file  
    with open(filename, 'r') as file:  
        e = json.load(file)  
    d = e[-1]  
  
    # add expectancy ratings of CS-  
    for i in range(len(d['expectancy_cs_minus'])):  
        if not d['expectancy_cs_minus'][i]:  
            addnum = np.nan  
        else:  
            addnum = float(d['expectancy_cs_minus'][i])  
        exp_minus.append(addnum)  
  
    # add expectancy ratings of CS+  
    for i in range(len(d['expectancy_cs_plus'])):  
        if not d['expectancy_cs_plus'][i]:  
            addnum = np.nan  
        else:  
            addnum = float(d['expectancy_cs_plus'][i])  
        exp_plus.append(addnum)  
  
    allData[sj]['exp_plus'] = np.array(exp_plus)  
    allData[sj]['exp_minus'] = np.array(exp_minus)  
  
    # save information on experiment design  
    StimSeq = []  
    for i in range(len(d['stimulus_image'])):  
        a = d['stimulus_image'][i]  
        if a[31] == 'c':  
            StimSeq.append(1)  
        else:  
            StimSeq.append(2)  
  
    us = []  
    for i in range(len(d['stimulus_audio'])):  
        a = d['stimulus_audio'][i]  
        if len(a) > 4 and a[33] == 's':  
            us.append(1)  
        else:  
            us.append(0)  
  
    exp_order = []  
    for i in range(len(d['trial_type'])):  
        a = d['trial_type'][i]  
        if a == 'fctask-trial':  
            exp_order.append(3)  
        elif a == 'expectancy-rating':  
            exp_order.append(4)  
  
    # save practice ratings  
    practRatings = []  
    for i in range(len(d['trial_type'])):  
        a = d['trial_type'][i]  
  
        if a == 'expectancy-rating':  
            newS = e[i]  
            if 'csRating' in newS:  
                try:  
                    practRatings.append(float(e[i]['csRating']))  
                except ValueError:  
                    practRatings.append(np.nan)  
  
    try:  
        j = 1  
        exp_order_reorg = exp_order.copy()  
        for k in range(len(exp_order)):  
            if exp_order[k] == 3:  
                exp_order_reorg[k] = StimSeq[j - 1]  
                j += 1  
  
        b = [i for i, x in enumerate(exp_order) if x == 3]  
    except Exception as e:  
        print(f'StimSeq not correct for subject {sj}')  
  
    # add affective ratings  
    for i in range(len(d['affective_cs_plus'])):  
        if not d['affective_cs_plus'][i]:  
            addnum = np.nan  
        else:  
            addnum = float(d['affective_cs_plus'][i])  
        pleasP.append(addnum)  
  
        if not d['affective_cs_minus'][i]:  
            addnum = np.nan  
        else:  
            addnum = float(d['affective_cs_minus'][i])  
        pleasM.append(addnum)  
  
    # save information for each subject in struct  
    allData[sj]['pleasM'] = pleasM  
    allData[sj]['pleasP'] = pleasP  
  
    allData[sj]['RTM'] = np.abs(d['fc_rt'][np.array(StimSeq) == 1])  
    allData[sj]['RTP'] = np.abs(d['fc_rt'][np.array(StimSeq) == 2])  
    allData[sj]['RT'] = np.abs(d['fc_rt'])  
  
    allData[sj]['StimSeq'] = StimSeq  
    allData[sj]['cs'] = StimSeq  
    allData[sj]['us'] = us  
    try:  
        allData[sj]['exp_order'] = exp_order  
    except Exception as e:  
        pass  
  
    allData[sj]['exp_order_reorg'] = exp_order_reorg  
    allData[sj]['ID'] = filename[0:24]  
    allData[sj]['d'] = d  
    allData[sj]['trial_type'] = d['trial_type']  
    allData[sj]['practRatings'] = practRatings  
  
exp_minus_mat = []  
exp_plus_mat = []  
pleasM_mat = []  
anxM_mat = []  
fearM_mat = []  
pleasP_mat = []  
anxP_mat = []  
fearP_mat = []  
RTM_mat = []  
RTP_mat = []  
RT_mat = []  
  
for sj in range(Nsj):  
    if len(allData[sj]['exp_minus']) == 20:  
        exp_minus_mat.append(allData[sj]['exp_minus'])  
        exp_plus_mat.append(allData[sj]['exp_plus'])  
    else:  
        print(f'sj {sj} not added')  
  
for sj in range(Nsj):  
    try:  
        if len(allData[sj]['pleasM']) == 6:  
            pleasM_mat.append(allData[sj]['pleasM'])  
            pleasP_mat.append(allData[sj]['pleasP'])  
  
        RTP_mat.append(allData[sj]['RTP'])  
        RTM_mat.append(allData[sj]['RTM'])  
        RT_mat.append(allData[sj]['RT'])  
  
    except Exception as e:  
        print(f'sj {sj} not added')  
  
exp_minus_mat = np.array(exp_minus_mat).T  
exp_plus_mat = np.array(exp_plus_mat).T  
pleasM_mat = np.array(pleasM_mat)  
pleasP_mat = np.array(pleasP_mat)  
RTM_mat = np.array(RTM_mat).T  
RTP_mat = np.array(RTP_mat).T  
RT_mat = np.array(RT_mat).T  
  
mExpMinus = np.mean(exp_minus_mat, axis=1)  
mExpPlus = np.mean(exp_plus_mat, axis=1)  
  
mPleasM = np.mean(pleasM_mat, axis=0)  
mPleasP = np.mean(pleasP_mat, axis=0)  
  
mRTP = np.mean(RTP_mat, axis=1)  
mRTM = np.mean(RTM_mat, axis=1)  
mRT = np.mean(RT_mat, axis=1)  
  
# prep data for plotting  
nTrials = (len(mRTM) + len(mRTP))  
  
CSplus = []  
CSminus = []  
otherCS = []  
for i in range(nTrials):  
    if StimSeq[i] == 2:  # double-check if StimEst is the same for everybody.  
        CSplus.append(mRT[i])  
        CSminus.append(np.nan)  
    else:  
        CSminus.append(mRT[i])  
        CSplus.append(np.nan)  
  
CSplusInt = np.interp(np.arange(len(CSplus)), np.flatnonzero(~np.isnan(CSplus)), np.array(CSplus)[~np.isnan(CSplus)])  
CSminusInt = np.interp(np.arange(len(CSminus)), np.flatnonzero(~np.isnan(CSminus)), np.array(CSminus)[~np.isnan(CSminus)])  
  
plt.plot(CSminusInt[4:], linewidth=2)  
plt.plot(CSplusInt[4:], linewidth=2)  
plt.show()  
  
# plot average expectancy ratings and reaction times  
fig, (ax1, ax2) = plt.subplots(1, 2)  
  
ax1.plot(range(1, 21), mExpMinus, linewidth=3)  
ax1.plot(range(1, 21), mExpPlus, linewidth=3)  
  
ax2.plot(range(1, 29), mRTM)  
ax2.plot(range(1, 35), mRTP)  
  
plt.show()  
  
# plot affective ratings  
fig, ax = plt.subplots()  
  
ax.plot(range(1, 7), mPleasM, linewidth=3)  
ax.plot(range(1, 7), mPleasP, linewidth=3)  
ax.set_ylim([1, 100])  
ax.set_ylabel('Pleasant')  
  
plt.show()  
